### PR TITLE
Implement SOM training and cluster assignment

### LINKED ISSUE #325: Create basic SOM implementation
**Is your feature request related to a problem? Please describe.**
Addresses part 1 of #287. We currently don't have a way to run SOMs in the repo. 

**Describe the solution you'd like**
As a first step to setting up an executable script for Candace to run, we should first create a default template for the repo. The functionality should be the following:

1. Load all images in a given directory into a numpy array
2. Use images to train a SOM, copying the same hyper-parameters that FlowSOM currently uses. 
3. Run SOM on all pixels in the image. 
4. Produce output file in a convenient form to allow comparison with existing FlowSOM clusters

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Addresses and closes #325. Addresses half of #345. We add the SOM training and testing processes in this PR.

**How did you implement your changes**

We'll be working off of the MiniSOM repo which provides a nice lightweight and intuitive implementation of SOMs. In line with the rest of our repo, we convert the process to a functional approach and only take the functions that are applicable. The default activation, neighborhood, and distance functions should correspond to the defaults for the R SOM implementation as well.

Our work will be in `cluster.py` and `cluster_test.py`.

**Remaining issues**

We try to use the same parameters that the R SOM package uses (which is what Candace calls it with), but there are a few things that we may not be able to duplicate completely (ex. the default functionality of R SOM is to also specify an end learning rate, whereas MiniSOM has no such requirement). Hopefully, this does not lead to significant differences in the clustering results.

While the original MiniSOM author seems to have done his best to speed up the process with optimized NumPy code, the biggest bottleneck right now seems to be the number of rows Candace has (almost 4 million, which means 40 million updates because R SOM defaults to 10 passes through the data). No joke: training the SOM on the entire dataset takes 3 HOURS to run. Thank goodness Tyler had some stuff to do 😅. While we will address subsetting in a future PR, for now, I think it would be best to ask Candace to run FlowSOM and return the output of SOM on a much smaller subset of her data. Using the same subset, we'll compare the outputs we get on our end. It will greatly speed up verification for this PR.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
Still need to add testing and potential optimizations in, but please let me know how this looks so far!

@ngreenwald using our minimized functional approach, I was able to perfectly replicate this example in the MiniSOM repo which most closely aligns with our goals: https://github.com/JustGlowing/minisom/blob/master/examples/Clustering.ipynb. So I think it's cool to begin writing the test functions assuming they work.

**Review by ngreenwald — APPROVED** [review]
This looks great. I think you should edit the description so that this only closes the first issue, and create a separate PR for implementing testing. Smaller PRs like this are much easier to review. 

For now, I don't think it's worth updating the example jupyter notebook. This is likely going to change substantially as we make more changes to the API/frontend of the SOM stuff. I'm sure it's useful for you to use to test stuff out, but I don't think these changes need to go in the PRs. 

I agree that eventually we want Candace to run the SOM on a smaller subset of the data. But the first step I think should be to run both methods on the entire dataset, since that's what she's doing. Now that you've reproduced one of the examples from the miniSOM repo, the next step is to write our own tests. Once those are done, I think we can make a version for candace to run, and compare the outputs

**Review by ackagel — APPROVED** [review]
Looks great!
